Compilation fails on GHC-9.4.1

Here is the log of how I tried to compile the code:

```
janus@gorm ~/flipstone % gh repo clone basvandijk/threads
Cloning into 'threads'...
remote: Enumerating objects: 469, done.
remote: Total 469 (delta 0), reused 0 (delta 0), pack-reused 469
Receiving objects: 100% (469/469), 95.95 KiB | 564.00 KiB/s, done.
Resolving deltas: 100% (208/208), done.
janus@gorm ~/flipstone % cd threads 
janus@gorm ~/flipstone/threads
 % cabal test -w ghc-9.4.1
Resolving dependencies...
Build profile: -w ghc-9.4.1 -O1
In order, the following will be built (use -v for more details):
 - concurrent-extra-0.7.0.12 (lib) (requires build)
 - hostname-1.0 (lib:hostname) (requires build)
 - xml-1.3.14 (lib:xml) (requires build)
 - test-framework-0.8.2.0 (lib) (requires build)
 - test-framework-hunit-0.3.0.2 (lib:test-framework-hunit) (requires build)
 - threads-0.5.1.6 (first run)
Starting     hostname-1.0 (all, legacy fallback)
Starting     xml-1.3.14 (all, legacy fallback)
Starting     concurrent-extra-0.7.0.12 (lib)
Building     xml-1.3.14 (all, legacy fallback)
Building     concurrent-extra-0.7.0.12 (lib)
Building     hostname-1.0 (all, legacy fallback)
Installing   hostname-1.0 (all, legacy fallback)
Completed    hostname-1.0 (all, legacy fallback)
Installing   concurrent-extra-0.7.0.12 (lib)
Completed    concurrent-extra-0.7.0.12 (lib)
Installing   xml-1.3.14 (all, legacy fallback)
Completed    xml-1.3.14 (all, legacy fallback)
Starting     test-framework-0.8.2.0 (lib)
Building     test-framework-0.8.2.0 (lib)
Installing   test-framework-0.8.2.0 (lib)
Completed    test-framework-0.8.2.0 (lib)
Starting     test-framework-hunit-0.3.0.2 (all, legacy fallback)
Building     test-framework-hunit-0.3.0.2 (all, legacy fallback)
Installing   test-framework-hunit-0.3.0.2 (all, legacy fallback)
Completed    test-framework-hunit-0.3.0.2 (all, legacy fallback)
[1 of 2] Compiling Main             ( /home/janus/flipstone/threads/dist-newstyle/build/x86_64-linux/ghc-9.4.1/threads-0.5.1.6/setup/setup.hs, /home/janus/flipstone/threads/dist-newstyle/build/x86_64-linux/ghc-9.4.1/threads-0.5.1.6/setup/Main.o )
[2 of 2] Linking /home/janus/flipstone/threads/dist-newstyle/build/x86_64-linux/ghc-9.4.1/threads-0.5.1.6/setup/setup
Configuring threads-0.5.1.6...
Warning: Packages using 'cabal-version: >= 1.10' and before 'cabal-version:
3.4' must specify the 'default-language' field for each component (e.g.
Haskell98 or Haskell2010). If a component uses different languages in
different modules then list the other ones in the 'other-languages' field.
Preprocessing library for threads-0.5.1.6..
Building library for threads-0.5.1.6..
[1 of 3] Compiling Control.Concurrent.Raw ( Control/Concurrent/Raw.hs, /home/janus/flipstone/threads/dist-newstyle/build/x86_64-linux/ghc-9.4.1/threads-0.5.1.6/build/Control/Concurrent/Raw.o, /home/janus/flipstone/threads/dist-newstyle/build/x86_64-linux/ghc-9.4.1/threads-0.5.1.6/build/Control/Concurrent/Raw.dyn_o )

Control/Concurrent/Raw.hs:16:16: error:
    • Couldn't match expected type: ghc-prim-0.9.0:GHC.Prim.State#
                                      ghc-prim-0.9.0:GHC.Prim.RealWorld
                                    -> (# ghc-prim-0.9.0:GHC.Prim.State#
                                            ghc-prim-0.9.0:GHC.Prim.RealWorld,
                                          a1 #)
                  with actual type: IO ()
    • In the first argument of ‘fork#’, namely ‘action’
      In the expression: fork# action s
      In the expression:
        case (fork# action s) of (# s1, tid #) -> (# s1, ThreadId tid #)
   |
16 |    case (fork# action s) of (# s1, tid #) -> (# s1, ThreadId tid #)
   |                ^^^^^^

Control/Concurrent/Raw.hs:21:22: error:
    • Couldn't match expected type: ghc-prim-0.9.0:GHC.Prim.State#
                                      ghc-prim-0.9.0:GHC.Prim.RealWorld
                                    -> (# ghc-prim-0.9.0:GHC.Prim.State#
                                            ghc-prim-0.9.0:GHC.Prim.RealWorld,
                                          a0 #)
                  with actual type: IO ()
    • In the second argument of ‘forkOn#’, namely ‘action’
      In the expression: forkOn# cpu action s
      In the expression:
        case (forkOn# cpu action s) of
          (# s1, tid #) -> (# s1, ThreadId tid #)
   |
21 |    case (forkOn# cpu action s) of (# s1, tid #) -> (# s1, ThreadId tid #)
```

--- a/Control/Concurrent/Raw.hs
+++ b/Control/Concurrent/Raw.hs
@@ -1,4 +1,4 @@
-{-# LANGUAGE NoImplicitPrelude, MagicHash, UnboxedTuples #-}
+{-# LANGUAGE CPP, NoImplicitPrelude, MagicHash, UnboxedTuples #-}
 
 module Control.Concurrent.Raw ( rawForkIO, rawForkOn ) where
 
@@ -12,10 +12,18 @@ import GHC.Conc      ( ThreadId(ThreadId) )
 -- exception handler.
 {-# INLINE rawForkIO #-}
 rawForkIO :: IO () -> IO ThreadId
+#if MIN_VERSION_base(4,17,0)
+rawForkIO (IO action) = IO $ \s ->
+#else
 rawForkIO action = IO $ \s ->
+#endif
    case (fork# action s) of (# s1, tid #) -> (# s1, ThreadId tid #)
 
 {-# INLINE rawForkOn #-}
 rawForkOn :: Int -> IO () -> IO ThreadId
+#if MIN_VERSION_base(4,17,0)
+rawForkOn (I# cpu) (IO action) = IO $ \s ->
+#else
 rawForkOn (I# cpu) action = IO $ \s ->
+#endif
    case (forkOn# cpu action s) of (# s1, tid #) -> (# s1, ThreadId tid #)

--- a/test/test.hs
+++ b/test/test.hs
@@ -205,10 +205,10 @@ wrapOn_0 :: (Fork a -> IO b) -> IO b
 wrapOn_0 = wrap $ ThreadGroup.forkOn 0
 
 wrapIOWithUnmask :: (Fork a -> IO b) -> IO b
-wrapIOWithUnmask = wrap $ \tg m -> ThreadGroup.forkIOWithUnmask tg $ const m
+wrapIOWithUnmask = wrap $ \tg m -> ThreadGroup.forkIOWithUnmask tg (\_ -> m)
 
 wrapOnWithUnmask :: (Fork a -> IO b) -> IO b
-wrapOnWithUnmask = wrap $ \tg m -> ThreadGroup.forkOnWithUnmask 0 tg $ const m
+wrapOnWithUnmask = wrap $ \tg m -> ThreadGroup.forkOnWithUnmask 0 tg (\_ -> m)
 
 wrap :: (ThreadGroup -> Fork a) -> (Fork a -> IO b) -> IO b
 wrap doFork test = ThreadGroup.new >>= test . doFork

